[clang] Improved isSimpleTypeSpecifier

carlos4242 opened the pull request:
- Sema::isSimpleTypeSpecifier return true for _Bool in c99 (currently returns false for _Bool, regardless of C dialect). (Fixes #72203)
- replace the logic with a check for simple types and a proper check for a valid keyword in the appropriate dialect

carlos4242 commented:
@owenca ... is this better for you?

owenca commented:
See https://discourse.llvm.org/t/should-issimpletypespecifier-return-true-for-bool/74873/16.

Commit:
[clang]
    - Sema::isSimpleTypeSpecifier return true for _Bool in c99 (currently returns false for _Bool, regardless of C dialect). (Fixes #72203)
    - replace the logic with a check for simple types and a proper check for a valid keyword in the appropriate dialect
    - PR feedback fixes for owenca

owenca commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -156,24 +161,19 @@ bool Sema::isSimpleTypeSpecifier(tok::TokenKind Kind) const {
-    return true;
-
-  case tok::annot_typename:
+  case tok::kw__Bool:
```
I would move it up, to just below `kw_bool`.

owenca commented on clang/include/clang/Sema/Sema.h:
```diff
@@ -2636,7 +2636,7 @@ class Sema final {
   void DiagnoseUseOfUnimplementedSelectors();
 
-  bool isSimpleTypeSpecifier(tok::TokenKind Kind) const;
+  bool isSimpleTypeSpecifier(Token &Tok) const;
```
```suggestion
  bool isSimpleTypeSpecifier(const Token &Tok) const;
```

owenca commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -128,10 +128,15 @@ class TypeNameValidatorCCC final : public CorrectionCandidateCallback {
 
 /// Determine whether the token kind starts a simple-type-specifier.
-bool Sema::isSimpleTypeSpecifier(tok::TokenKind Kind) const {
+bool Sema::isSimpleTypeSpecifier(Token &Tok) const {
+  auto Kind = Tok.getKind();
+  auto LangOpts = getLangOpts();
+
   switch (Kind) {
```
```suggestion
bool Sema::isSimpleTypeSpecifier(const Token &Tok) const {
  switch (Tok.getKind()) {
```

owenca commented on clang/lib/Sema/SemaDecl.cpp:
```diff
@@ -156,24 +161,19 @@ bool Sema::isSimpleTypeSpecifier(tok::TokenKind Kind) const {
-
   case tok::kw_char8_t:
-    return getLangOpts().Char8;
+    if (!Tok.hasIdentifierInfo())
+      return false;
+    return Tok.getIdentifierInfo()->isKeyword(LangOpts);
```
```suggestion
    return Tok.getIdentifierInfo()->isKeyword(getLangOpts());
```

owenca commented on clang/include/clang/Lex/Token.h:
```diff
@@ -196,6 +196,13 @@ class Token {
     PtrData = (void*) II;
   }
 
+  bool hasIdentifierInfo() {
+    if (is(tok::raw_identifier) || isAnnotation() || isLiteral() ||
+        is(tok::eof))
+      return false;
+    return true;
+  }
+
```
IMO we don't need this.

carlos4242 commented:
I think unit testing the bug fix to this function (the copy that is in Sema) is going to be nearly impossible.

To prove the point, I tried hard coding the wrong return value for many types just now...

```
  switch (Kind) {
  case tok::kw_bool:
  case tok::kw_short:
  case tok::kw_long:
  case tok::kw___int64:
  case tok::kw___int128:
  case tok::kw_signed:
  case tok::kw_unsigned:
  case tok::kw_char:
  case tok::kw_int:
  case tok::kw_half:
  case tok::kw_float:
  case tok::kw_double:
  case tok::kw___bf16:
  case tok::kw__Float16:
  case tok::kw___float128:
  case tok::kw___ibm128:
  case tok::kw_wchar_t:
    return false;
```

...Then re-ran the full clang/tests suite (18,000+ tests) and all passed as normal. So there's basically very little unit test coverage on this function as it stands. It's not just a question of improving or adding to existing tests. Whoever first wrote this function seems to have not added really enough unit test coverage at the time, and it's only called in two obscure places in Sema. The duplicate copy function in Format is more widely used and would probably be easier to test. But for this PR I think I'll make the changes you suggest, then leave it as is... either it gets approved without new tests or it'll get culled one day as an abandoned PR. I'll leave that up to the maintainers.

owenca commented:
LGTM, but please rebase and resolve conflicts.

carlos4242 commented:
> LGTM, but please rebase and resolve conflicts.

done, rebased, tests pass locally

carlos4242 commented:
note: build kite is passing on linux but failing windows tests with an error like "mt.exe : general error c101008d: Failed to write the updated manifest to the resource of file "bin\llvm-symbolizer.exe". Operation did not complete successfully because the file contains a virus or potentially unwanted software" ... I'm a bit unsure how to proceed with that. Does it look like a genuine failure I should address or a spurious build quirk?

carlos4242 committed:
Merge branch 'main' into improved-isSimpleTypeSpecifier2

AaronBallman reviewed (approved):
LGTM; I can't see a reasonable way for us to test this change and it's *almost* an NFC change, so I think it's fine to land without tests given that we're at the very start of the 19.x cycle.

AaronBallman commented:
Once precommit CI comes back green, I'll merge on your behalf.
